### DISCUSSION TITLE
4/5 PWM channels but only 3 as a Color led the rest separately?

### DISCUSSION THREAD

**Discussion by msperl** [arendst/Tasmota#17995] [Q&A]
Is there a way to limit the amount of LED that are used by the lights integration?

THe question I am asking is becasue with a Lanbon L8 Diesplay I have 3 PWM channels for a RGB color strip at the bezel and 1 PWM channel for the Backlight.

I know there is:
- SetOption68 to separate the RGB into separate channels if wanted (but that is no longer documented and does nto do the full monty either...)
- SetOption37 to decide on the order of RGB mapped to PWM channels - but that one tries to take all the PWM channels - and that includes my backlight channel as well.

Both options give strange effects in HomeAssistant integration as well and do not map really well.

Possibly part of the problem is that at least with LVGL I can not really controll the Backlight (or I have not found a way to contol the backlight) when configuring GPIO5 as backlight - subsequently the sceen stays dark unless I configure the Pin 5 as PWM channel 4.... But then it is interpreted as Cold White and mixed into the sceeen setting.

To put it into perspective:
Here the way it looks in the TASMOTA UI:
<img width="367" alt="image" src="https://user-images.githubusercontent.com/2638784/219959917-7f3bf064-00e9-4140-8034-37a136e5c9a0.png">

And in home assistant it looks like this when freshly discovered (same settings as in the Tasmota UI):
<img width="667" alt="image" src="https://user-images.githubusercontent.com/2638784/219959873-d34ee3fa-0646-4cfe-a5a7-e3599d9f1225.png">

Here the identified version:
<img width="667" alt="image" src="https://user-images.githubusercontent.com/2638784/219959792-0026f195-1f30-4190-a1c8-20d3646ac5ac.png">

(for reproducability I had to run things backward - Identified HA, deleted device, repowered device readded to HA and finally the WebUI screenshot) 

So you see:
- led/on/off on its own
- Backlite brightness picker mixed with ScreenClear OnOff
- Led color picker mixed with Backlite On/Off (this has some strange semantics after a device restart)

For completeness here the State message taken from the console:
```
17:09:22.233 MQT: tele/nit-d-test/STATE = {"Time":"2023-02-19T17:09:22","Uptime":"0T00:04:41","UptimeSec":281,"Heap":47,"SleepMode":"Dynamic","Sleep":10,"LoadAvg":102,"MqttCount":1,"Berry":{"HeapUsed":30,"Objects":384},"POWER1":"OFF","POWER2":"OFF","POWER3":"OFF","POWER4":"OFF","Dimmer1":85,"POWER5":"ON","Dimmer2":18,"Color":"00D8002E","HSBColor":"120,100,85","White":18,"Channel":[0,85,0,18],"Scheme":0,"Fade":"OFF","Speed":1,"LedTable":"ON","Wifi":{"AP":1,"SSId":...}}
```

Here the used template with Backlite as PWM4 and GPIO15 as  Option A3:
```
{"NAME":"Lanbon L8","GPIO":[0,0,0,0,0,419,0,0,224,0,225,6210,0,0,1024,896,0,960,928,864,0,832,416,226,0,0,0,0,417,418,0,352,0,0,0,4736],"FLAG":0,"BASE":1}
```
And for comparison the one mentioned in by blakadder.com:
```
{"NAME":"Lanbon L8","GPIO":[0,0,0,0,0,992,0,0,224,0,225,0,0,0,1024,896,0,6624,6592,864,0,832,416,226,0,0,0,0,417,418,0,352,0,0,0,4736],"FLAG":0,"BASE":1}
```

The display.ini is taken from the repo with the change that backlight is not assigned as PIN 5 but -1:
```
:H,ST7789,240,320,16,SPI,1,22,19,23,21,-1,18,25,40
```

any ideas how to improve the situation in HomeAssistant in the end?
(Using Tasmota release 12.4.0 build from GIT checkout)

**Comment by barbudor** [comment]
You can't have 2 "lights" in Tasmota
As you are on ESP32 you have Berry that can come to help
I suggest that you keep the led strip as a light and then use Berry to manage the backlight of the lcd. For that, don't define the GPIO for the backlight PWM in the template but directly in Berry.

**Reply by msperl** [reply]
How would I do that in Berry?
And how would I expose it then to Home Assistant, so that I can control the screen brightness from there?

**Reply by msperl** [reply]
ok  for PWM in principle I would use: gpio.set_pwm.
Understood so far - but how can i integrate it into the HA integration to get both working?
Thanks,

**Reply by msperl** [reply]
OK - so in principle with this I get Backlight control:
```
import gpio
gpio.pin_mode(5,gpio.OUTPUT)
gpio.set_pwm(5,100)
```

But with a side-note that running the pin_mode multiple times makes the display go dark.
Running only `gpio.set_pwm` works and allows changing the level.
SImilarly after a gpio.digital_write a gpio.set_pwm does not work 
(or at least the naive way I have tested it)

**Reply by barbudor** [reply]
I would very question the interrest of reporting the backlight to HA 😄 
But you can publish on topics that would fit and do a manual integration in HA (but as a non-user of HA, I can't give you more directions)

**Reply by msperl** [reply]
My use case: modifying brightness of the display from a Mobile device via HomeAssistant UI.
And ideally modifying the color of the RGB leds separately - so that is 2 separate lights (and as far as I understand at least berry light.get supports multiple lights)

But in the meantime I have come accross something strange, that may be messing with things in the first place:
I run with the following minimal template
```
{"NAME":"test","GPIO":[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,1,1,1,0,1,1,1,0,0,0,0,1,1,6210,1,1,0,0,1],"FLAG":0,"BASE":1}
```
Only Option A3 set on GPIO 34.

And On the Tasmnota WebUI there is all of a sudden an On/Off Button!
Is this expected that a UniversalDisplay adds a switch?

```
19:09:37.400 RSL: STATE = {"Time":"2023-02-19T19:09:37","Uptime":"0T00:10:08","UptimeSec":608,"Heap":50,"SleepMode":"Dynamic","Sleep":10,"LoadAvg":99,"MqttCount":0,"Berry":{"HeapUsed":29,"Objects":376},"POWER":"ON","Wifi":{}}
```

Maybe this is the thing messing with the lights - at least in the export to HA?

The other thing is that when I now modify the GPIO via Berry:
```
import gpio
gpio.pin_mode(5,gpio.OUTPUT)
gpio.set_pwm(5,100)
```

I also modify the RED LED brightness - as if the primary PWM mapped to Pin 26 as well...
I need to do a gpio.set_pwm(26,0) to make the RED LED to get controlled separately!

**Comment by msperl** [comment]
Actually I have had a look at the documentation: here: https://tasmota.github.io/docs/Lights/#pwm6
```
ESP32 has hardware PWM support, named ledc, for up to 16 channels depending on CPU type. You can mix lights and pure PWM channels. The first 5 PWM are reserved for lights, unless SetOption15 0. For pure PWM GPIOs, you can assign any PWM number, they don't need to be continuous. For example you can use PWM 1/2/3 for a 3-channel RGB light, and PWM 6 & PWM 10 for pure PWM at the same time.
```

On the other hand with berry docu one can find: https://tasmota.github.io/docs/Berry/#light-object
```
light.get | (index:int) -> mapGet the current status if light number index (default:0).
```

So in principle: both indicate if multiple lights would be supported.

On the other hand on the web-ui with GPIO5 set as PWM 6 (So PWM1-3,6 configured) I only find the RGB light.
So it does not seem to be looking for PWM when there are gaps of PWM that are unused.

as for controlling the LVGL Display brightness this is also working as soon as the backlit pin is configured in the display description:
```
import display
display.dimmer(10)
```

So now it looks like this with GPIO5 configured as backlight:
```
09:39:45.732 MQT: tele/nit-d-test/STATE = {"Time":"2023-02-20T09:39:45","Uptime":"0T00:05:08","UptimeSec":308,"Heap":51,"SleepMode":"Dynamic","Sleep":10,"LoadAvg":99,"MqttCount":1,"Berry":{"HeapUsed":29,"Objects":377},"POWER1":"OFF","POWER2":"OFF","POWER3":"OFF","POWER4":"OFF","Dimmer":100,"Color":"FFFFFF","HSBColor":"0,0,100","Channel":[100,100,100],"Scheme":0,"Fade":"OFF","Speed":1,"LedTable":"ON","Wifi":{...}}
09:45:25.396 CMD: template
09:45:25.406 MQT: stat/nit-d-test/RESULT = {"NAME":"Lanbon L8","GPIO":[0,0,0,0,0,992,0,0,224,0,225,0,0,0,1024,896,0,960,928,864,0,832,416,226,0,0,0,0,417,418,6210,352,0,0,0,4738],"FLAG":0,"BASE":1}
```

if I change GPIO 5 in the template from backlight to PWM6:
```
10:01:35.436 CMD: template
10:01:35.447 MQT: stat/nit-d-test/RESULT = {"NAME":"Lanbon L8","GPIO":[0,0,0,0,0,421,0,0,224,0,225,0,0,0,1024,896,0,960,928,864,0,832,416,226,0,0,0,0,417,418,6210,352,0,0,0,4738],"FLAG":0,"BASE":1}
10:01:38.572 MQT: tele/nit-d-test/STATE = {"Time":"2023-02-20T10:01:38","Uptime":"0T00:02:09","UptimeSec":129,"Heap":51,"SleepMode":"Dynamic","Sleep":10,"LoadAvg":99,"MqttCount":1,"Berry":{"HeapUsed":29,"Objects":377},"POWER1":"OFF","POWER2":"OFF","POWER3":"OFF","POWER4":"OFF","Dimmer":100,"Color":"FFFFFF","HSBColor":"0,0,100","Channel":[100,100,100],"Scheme":0,"Fade":"OFF","Speed":1,"LedTable":"ON","PWM":{"PWM1":0,"PWM2":0,"PWM3":0,"PWM6":0},"Wifi":{...}}
```
So now that backlight=PWM6 PWM is now reported - at least as PWM not as a light (and not visible in the Web UI)

Also PWM is reported as 0 even though PWM is set to 50 via:
```
import display
display.dimmer(50)
```

So in the end the display.dimmer approach is also not very helpfull - and it does not expose the value over MQTT either - not even read only!

looks like this (including the message when changing the color via HA):
```
10:38:53.299 MQT: stat/nit-d-test/RESULT = {"POWER4":"ON","Dimmer":60,"Color":"021199","HSBColor":"234,99,60","Channel":[1,6,60]}
10:39:08.569 MQT: tele/nit-d-test/STATE = {"Time":"2023-02-20T10:39:08","Uptime":"0T00:39:39","UptimeSec":2379,"Heap":50,"SleepMode":"Dynamic","Sleep":10,"LoadAvg":99,"MqttCount":1,"Berry":{"HeapUsed":29,"Objects":377},"POWER1":"OFF","POWER2":"OFF","POWER3":"OFF","POWER4":"ON","Dimmer":60,"Color":"021199","HSBColor":"234,99,60","Channel":[1,6,60],"Scheme":0,"Fade":"OFF","Speed":1,"LedTable":"ON","PWM":{"PWM1":8,"PWM2":32,"PWM3":329,"PWM6":0},"Wifi":{...}}
```
So we see color, we see PWM for the light, but we do not see PWM for the backlite one!

There are all sorts of inconsistencies that hint different things and I am not sure what the vision should be here.

In the end I want to know what ist the "sanctioned" approach to set up the display backlite, so that it does not conflict with colors leds and is ideallly managable via the UI...

**Reply by s-hadinger** [reply]
That's the beauty of legacy and backwards compatibility... PWM1-5 are reserved for lights. Backlight is supposed to be set as `Backlight` and controlled with `DisplayDimmer` command. There is no UI for backlight dimmer value.

**Reply by msperl** [reply]
One Idea could have been another SetOptionX, that would support the PWMS 6-15 as additional lights (say say in different variants)
And then expose one of those - so backwards compatiblity would not be involved...

Something like that could probably get done with Berry, but the question is: how would it integrate seamlessly into the UI and possibly get picked up automatically by HA and other integrations - probably some work is needed there as well...

So from a Command perspective something like this could handle additional lights:
```
LightConfig<id> <pwm R>,<pwm G>,<pwm B>,<pwm CW>,<pwm WW>
```

where the channel could be "none", 1-16 for the corresponding PWM and "backlight" for the backlight...

Where LightConfig1 would be a more flexible way than SetOption37 (which is limited to PWM 1-5) - the  bit 7 versions would not be implemented, as they would get implemented as separate lights - so in the end ou could have 5 RGB lightstrips, 3 RGBWW or up to 16 single color strips - to some extend a flag to disable the old LED code would be needed, so that the PWM channels would not get highjacked...

Obviously Hooks into MQTT, the WebUI would also be needed (never done that so far)...

If there is interest I might give that a try with Berry...
